File: server/board.py
HORIZONTAL = 0
VERTICAL = 1

class Corner:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.occupied = False
        self.wall = None
    def __str__(self):
        return "Corner(x=" + str(self.x) + ",y=" + str(self.y) + ",occupied=" + str(self.occupied) + ")"

class Tile:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.up = True
        self.down = True
        self.left = True
        self.right = True
        self.topleft = None
        self.topright = None
        self.bottomleft = None
        self.bottomright = None
        self.reachable = False
    def __str__(self):
        return "Tile(x=" + str(self.x) + ",y=" + str(self.y) + ",up=" + str(self.up) + ",down=" + str(self.down) + ",left=" + str(self.left) + ",right=" + str(self.right) + ")"

class Wall:
    def __init__(self, x, y, direction):
        self.x = x
        self.y = y
        self.direction = direction
        self.owner = None

class Pawn:
    def __init__(self, num, board, x, y, startingwalls):
        self.x = x
        self.y = y
        self.num = num
        self.board = board
        self.wallsleft = startingwalls
    def __str__(self):
        return "Pawn" + str(self.num) + "(" + str(self.x) + "," + str(self.y) + ") with " + str(self.wallsleft) + " walls left"
    def isFinished(self):
        if self == self.board.pawn1:
            if self.y == 8:
                return True
            else:
                return False
        elif self == self.board.pawn2:
            if self.y == 0:
                return True
            else:
                return False
        else:
            print('invalid pawn')
            return False
    def reachNewTile(self, cur): #cur is a tile
        if cur.reachable == True:
            return
        x = cur.x
        y = cur.y
        cur.reachable = True
        if cur.up:
            self.reachNewTile(self.board.tiles[x][y+1])
        if cur.down:
            self.reachNewTile(self.board.tiles[x][y-1])
        if cur.left:
            self.reachNewTile(self.board.tiles[x-1][y])
        if cur.right:
            self.reachNewTile(self.board.tiles[x+1][y])
    def canReachFinishLine(self):
        for i in range(9):
            for j in range(9):
                self.board.tiles[i][j].reachable = False
        self.reachNewTile(self.board.tiles[self.x][self.y])
        result = False
        if self == self.board.pawn1:
            for i in range(9):
                if self.board.tiles[i][8].reachable:
                    result = True
        else:
            for i in range(9):
                if self.board.tiles[i][0].reachable:
                    result = True
        return result

    def move(self,x,y):
        enemy = self.board.pawn2 if self.board.pawn1 == self else self.board.pawn1
        dist = abs(self.x - x) + abs(self.y - y)
        if x<0 or y<0 or x>8 or y>8:
            return False
        if dist > 2:
            return False
        if dist == 1:
            currenttile = self.board.tiles[self.x][self.y]
            if x == self.x + 1 and currenttile.right and not (x == enemy.x and y == enemy.y):
                self.x = x
                return True
            elif x == self.x - 1 and currenttile.left and not (x == enemy.x and y == enemy.y):
                self.x = x
                return True
            elif y == self.y + 1 and currenttile.up and not (x == enemy.x and y == enemy.y):
                self.y = y
                return True
            elif y == self.y - 1 and currenttile.down and not (x == enemy.x and y == enemy.y):
                self.y = y
                return True
            return False
        if dist == 2:
            if y == self.y + 2:
                if enemy.x == self.x and enemy.y == self.y+1 and self.board.tiles[self.x][self.y].up and self.board.tiles[self.x][self.y+1].up:
                    self.y = y
                    return True
                else:
                    return False
            if y == self.y - 2:
                if enemy.x == self.x and enemy.y == self.y-1 and self.board.tiles[self.x][self.y].down and self.board.tiles[self.x][self.y-1].down:
                    self.y = y
                    return True
                else:
                    return False
            if x == self.x + 2:
                if enemy.x == self.x+1 and enemy.y == self.y and self.board.tiles[self.x][self.y].right and self.board.tiles[self.x+1][self.y].right:
                    self.x = x
                    return True
                else:
                    return False
            if x == self.x - 2:
                if enemy.x == self.x-1 and enemy.y == self.y and self.board.tiles[self.x][self.y].left and self.board.tiles[self.x-1][self.y].left:
                    self.x = x
                    return True
                else:
                    return False
            if x == self.x + 1 and y == self.y + 1:
                if enemy.x == self.x+1 and enemy.y == self.y and self.board.tiles[self.x][self.y].right and self.board.tiles[self.x+1][self.y].up and not self.board.tiles[self.x+1][self.y].right:
                    self.x = x
                    self.y = y
                    return True
                elif enemy.x == self.x and enemy.y == self.y+1 and self.board.tiles[self.x][self.y].up and self.board.tiles[self.x][self.y+1].right and not self.board.tiles[self.x][self.y+1].up:
                    self.x = x
                    self.y = y
                    return True
                else:
                    return False
            if x == self.x + 1 and y == self.y - 1:
                if enemy.x == self.x+1 and enemy.y == self.y and self.board.tiles[self.x][self.y].right and self.board.tiles[self.x+1][self.y].down and not self.board.tiles[self.x+1][self.y].right:
                    self.x = x
                    self.y = y
                    return True
                elif enemy.x == self.x and enemy.y == self.y-1 and self.board.tiles[self.x][self.y].down and self.board.tiles[self.x][self.y-1].right and not self.board.tiles[self.x][self.y-1].down:
                    self.x = x
                    self.y = y
                    return True
                else:
                    return False
            if x == self.x - 1 and y == self.y - 1:
                if enemy.x == self.x-1 and enemy.y == self.y and self.board.tiles[self.x][self.y].left and self.board.tiles[self.x-1][self.y].down and not self.board.tiles[self.x-1][self.y].left:
                    self.x = x
                    self.y = y
                    return True
                elif enemy.x == self.x and enemy.y == self.y-1 and self.board.tiles[self.x][self.y].down and self.board.tiles[self.x][self.y-1].left and not self.board.tiles[self.x][self.y-1].down:
                    self.x = x
                    self.y = y
                    return True
                else:
                    return False
            if x == self.x - 1 and y == self.y + 1:
                if enemy.x == self.x-1 and enemy.y == self.y and self.board.tiles[self.x][self.y].left and self.board.tiles[self.x-1][self.y].up and not self.board.tiles[self.x-1][self.y].left:
                    self.x = x
                    self.y = y
                    return True
                elif enemy.x == self.x and enemy.y == self.y+1 and self.board.tiles[self.x][self.y].up and self.board.tiles[self.x][self.y+1].left and not self.board.tiles[self.x][self.y+1].up:
                    self.x = x
                    self.y = y
                    return True
                else:
                    return False
        return False

    def placeWall(self,x,y,direction):
        if self.wallsleft < 1:
            return False
        if x<0 or y<0 or x>7 or y>7:
            return False
        if self.board.corners[x][y].occupied:
            return False
        if direction == HORIZONTAL and self.board.tiles[x][y].up and self.board.tiles[x+1][y].up:
            wall = Wall(x,y,direction)
            wall.owner = self
            self.board.corners[x][y].occupied = True
            self.board.corners[x][y].wall = wall
            self.board.tiles[x][y].up = False
            self.board.tiles[x+1][y].up = False
            self.board.tiles[x][y+1].down = False
            self.board.tiles[x+1][y+1].down = False
            self.wallsleft-=1
            if not self.board.pawnsCanReachFinishLine():
                self.removeWall(x,y)
                return False
            return True
        if direction == VERTICAL and self.board.tiles[x][y].right and self.board.tiles[x][y+1].right:
            wall = Wall(x,y,direction)
            wall.owner = self
            self.board.corners[x][y].occupied = True
            self.board.corners[x][y].wall = wall
            self.board.tiles[x][y].right = False
            self.board.tiles[x+1][y].left = False
            self.board.tiles[x][y+1].right = False
            self.board.tiles[x+1][y+1].left = False
            self.wallsleft-=1
            if not self.board.pawnsCanReachFinishLine():
                self.removeWall(x,y)
                return False
            return True
        return False

    def removeWall(self,x,y):
        if x<0 or y<0 or x>7 or y>7:
            return False
        if not self.board.corners[x][y].occupied:
            return False
        if self.board.corners[x][y].wall.direction == HORIZONTAL:
            owner = self.board.corners[x][y].wall.owner
            owner.wallsleft += 1
            self.board.corners[x][y].occupied = False
            self.board.corners[x][y].wall = None
            self.board.tiles[x][y].up = True
            self.board.tiles[x+1][y].up = True
            self.board.tiles[x][y+1].down = True
            self.board.tiles[x+1][y+1].down = True
            return True
        if self.board.corners[x][y].wall.direction == VERTICAL:
            owner = self.board.corners[x][y].wall.owner
            owner.wallsleft += 1
            self.board.corners[x][y].occupied = False
            self.board.corners[x][y].wall = None
            self.board.tiles[x][y].right = True
            self.board.tiles[x+1][y].left = True
            self.board.tiles[x][y+1].right = True
            self.board.tiles[x+1][y+1].left = True
            return True
        return False

class Board:
    def __init__(self):
        self.corners = []
        for i in range(8):
            self.corners.append([])
            for j in range(8):
                self.corners[i].append(Corner(i,j))
        self.tiles = []
        for i in range(9):
            self.tiles.append([])
            for j in range(9):
                _tile = Tile(i,j)
                if i==0:
                    _tile.left = False
                if i==8:
                    _tile.right = False
                if j==0:
                    _tile.down = False
                if j==8:
                    _tile.up = False
                if i>0 and j>0:
                    _tile.bottomleft = self.corners[i-1][j-1]
                if i<8 and j>0:
                    _tile.bottomright = self.corners[i][j-1]
                if i>0 and j<8:
                    _tile.topleft = self.corners[i-1][j]
                if i<8 and j<8:
                    _tile.topright = self.corners[i][j]
                self.tiles[i].append(_tile)
        self.pawn1 = Pawn(0,self,4,0,10)
        self.pawn2 = Pawn(1,self,4,8,10)
    def pawnsCanReachFinishLine(self):
        return self.pawn1.canReachFinishLine() and self.pawn2.canReachFinishLine()
    def __str__(self):
        res = []
        for i in range(17):
            res.append([])
            for j in range(17):
                res[i].append(" ")
            res[i].append("\n")

        for i in range(9):
            for j in range(9):
                if self.pawn1.x == i and self.pawn1.y == j:
                    res[16-j * 2][i * 2] = "▲"
                elif self.pawn2.x == i and self.pawn2.y == j:
                    res[16-j * 2][i * 2] = "▼"
                else:
                    res[16-j * 2][i * 2] = "□"
                if j<8 and not self.tiles[i][j].up:
                    res[16-(j*2+1)][i*2] = "━"
                if i<8 and not self.tiles[i][j].right:
                    res[16-j*2][i*2+1] = "┃"

        for i in range(8):
            for j in range(8):
                if self.corners[i][j].occupied:
                    if self.corners[i][j].wall.direction == HORIZONTAL:
                        res[16-(j*2+1)][i*2+1] = "━"
                    elif self.corners[i][j].wall.direction == VERTICAL:
                        res[16-(j*2+1)][i*2+1] = "┃"

        return "[player2] " + str(self.pawn2.wallsleft) + " walls left\n" + "".join(sum(res,[])) + "[player1] " + str(self.pawn1.wallsleft) + " walls left\n"

if __name__ == "__main__":
    board = Board()
    board.pawn1.move(4, 1)
    board.pawn2.placeWall(1, 1, HORIZONTAL)
    board.pawn2.placeWall(3, 1, HORIZONTAL)
    board.pawn1.move(5, 1)
    board.pawn2.placeWall(5, 1, HORIZONTAL)
    board.pawn2.placeWall(7, 1, HORIZONTAL)
    board.pawn1.placeWall(0, 1, VERTICAL)
    print(board)

File: webserver/static/board.js
const HORIZONTAL = 0;
const VERTICAL = 1;

class Corner{
    constructor(x,y){
        this.x = x;
        this.y = y;
        this.occupied = false;
        this.wall = null;
    }
}
class Tile{
    constructor(x, y){
        this.x = x;
        this.y = y;
        this.up = true;
        this.down = true;
        this.left = true;
        this.right = true;
        this.topleft = null;
        this.topright = null;
        this.bottomleft = null;
        this.bottomright = null;
        this.reachable = false;
    }
}
class Wall{
    constructor(x, y, direction){
        this.x = x;
        this.y = y;
        this.direction = direction;
        this.owner = null;
    }
}
class Pawn{
    constructor(num, board, x, y, startingwalls){
        this.x=x;
        this.y=y;
        this.num= num;
        this.board= board;
        this.wallsleft=startingwalls;
    }
    reachNewTile(cur){
        if(cur.reachable == true) return;
        var x=cur.x;
        var y=cur.y;
        cur.reachable=true;
        if(cur.up) this.reachNewTile(this.board.tiles[x][y+1]);
        if(cur.down) this.reachNewTile(this.board.tiles[x][y-1]);
        if(cur.left) this.reachNewTile(this.board.tiles[x-1][y]);
        if(cur.right) this.reachNewTile(this.board.tiles[x+1][y]);
    }
    canReachFinishLine(){
        for(var i=0;i<9;i++){
            for(var j=0;j<9;j++){
                this.board.tiles[i][j].reachable = false;
            }
        }
        this.reachNewTile(this.board.tiles[this.x][this.y]);
        var result = false;
        if(this == this.board.pawn1){
            for(var i=0;i<9;i++){
                if(this.board.tiles[i][8].reachable) result = true;
            }
        }
        else{
            for(var i=0;i<9;i++){
                if(this.board.tiles[i][0].reachable) result = true;
            }
        }
        return result;
    }
    move(x,y){
        var enemy = this.board.pawn1;
        if(this.board.pawn1 == this) enemy = this.board.pawn2;
        var dist = Math.abs(this.x - x) + Math.abs(this.y - y);
        if(x<0 || y<0 || x>8 || y>8) return false;
        if(dist > 2) return false;
        if(dist == 1){
            var currenttile = this.board.tiles[this.x][this.y];
            if(x == this.x + 1 && currenttile.right && !(x == enemy.x && y == enemy.y)){
                this.x = x;
                return true;
            }
            else if(x == this.x - 1 && currenttile.left && !(x == enemy.x && y == enemy.y)){
                this.x = x;
                return true;
            }
            else if(y == this.y + 1 && currenttile.up && !(x == enemy.x && y == enemy.y)){
                this.y = y;
                return true;
            }
            else if(y == this.y - 1 && currenttile.down && !(x == enemy.x && y == enemy.y)){
                this.y = y;
                return true;
            }
            return false;
        }
        if(dist == 2){
            if(y == this.y + 2){
                if(enemy.x == this.x && enemy.y == this.y+1 && this.board.tiles[this.x][this.y].up && this.board.tiles[this.x][this.y+1].up){
                    this.y = y;
                    return true;
                }
                else return false;
            }
            if(y == this.y - 2){
                if(enemy.x == this.x && enemy.y == this.y-1 && this.board.tiles[this.x][this.y].down && this.board.tiles[this.x][this.y-1].down){
                    this.y = y;
                    return true;
                }
                else return false;
            }
            if(x == this.x + 2){
                if(enemy.x == this.x+1 && enemy.y == this.y && this.board.tiles[this.x][this.y].right && this.board.tiles[this.x+1][this.y].right){
                    this.x = x;
                    return true;
                }
                else return false;
            }
            if(x == this.x - 2){
                if(enemy.x == this.x-1 && enemy.y == this.y && this.board.tiles[this.x][this.y].left && this.board.tiles[this.x-1][this.y].left){
                    this.x = x;
                    return true;
                }
                else return false;
            }
            if(x == this.x + 1 && y == this.y + 1){
                if(enemy.x == this.x+1 && enemy.y == this.y && this.board.tiles[this.x][this.y].right && this.board.tiles[this.x+1][this.y].up && !this.board.tiles[this.x+1][this.y].right){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else if(enemy.x == this.x && enemy.y == this.y+1 && this.board.tiles[this.x][this.y].up && this.board.tiles[this.x][this.y+1].right && !this.board.tiles[this.x][this.y+1].up){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else return false;
            }
            if(x == this.x + 1 && y == this.y - 1){
                if(enemy.x == this.x+1 && enemy.y == this.y && this.board.tiles[this.x][this.y].right && this.board.tiles[this.x+1][this.y].down && !this.board.tiles[this.x+1][this.y].right){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else if(enemy.x == this.x && enemy.y == this.y-1 && this.board.tiles[this.x][this.y].down && this.board.tiles[this.x][this.y-1].right && !this.board.tiles[this.x][this.y-1].down){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else return false;
            }
            if(x == this.x - 1 && y == this.y - 1){
                if(enemy.x == this.x-1 && enemy.y == this.y && this.board.tiles[this.x][this.y].left && this.board.tiles[this.x-1][this.y].down && !this.board.tiles[this.x-1][this.y].left){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else if(enemy.x == this.x && enemy.y == this.y-1 && this.board.tiles[this.x][this.y].down && this.board.tiles[this.x][this.y-1].left && !this.board.tiles[this.x][this.y-1].down){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else return false;
            }
            if(x == this.x - 1 && y == this.y + 1){
                if(enemy.x == this.x-1 && enemy.y == this.y && this.board.tiles[this.x][this.y].left && this.board.tiles[this.x-1][this.y].up && !this.board.tiles[this.x-1][this.y].left){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else if(enemy.x == this.x && enemy.y == this.y+1 && this.board.tiles[this.x][this.y].up && this.board.tiles[this.x][this.y+1].left && !this.board.tiles[this.x][this.y+1].up){
                    this.x = x;
                    this.y = y;
                    return true;
                }
                else return false;
            }
        }
        return false;
    }
    placeWall(x,y,direction){
        if(this.wallsleft < 1) return false;
        if(x<0 || y<0 || x>7 || y>7) return false;
        if(this.board.corners[x][y].occupied) return false;
        if(direction == HORIZONTAL && this.board.tiles[x][y].up && this.board.tiles[x+1][y].up){
            var wall = new Wall(x,y,direction);
            wall.owner = this;
            this.board.corners[x][y].occupied = true;
            this.board.corners[x][y].wall = wall;
            this.board.tiles[x][y].up = false;
            this.board.tiles[x+1][y].up = false;
            this.board.tiles[x][y+1].down = false;
            this.board.tiles[x+1][y+1].down = false;
            this.wallsleft-=1;
            if(!this.board.pawnsCanReachFinishLine()){
                this.removeWall(x,y);
                return false;
            }
            console.log('%d walls left', this.wallsleft);
            return true;
        }
        if(direction == VERTICAL && this.board.tiles[x][y].right && this.board.tiles[x][y+1].right){
            var wall = new Wall(x,y,direction);
            wall.owner = this;
            this.board.corners[x][y].occupied = true;
            this.board.corners[x][y].wall = wall;
            this.board.tiles[x][y].right = false;
            this.board.tiles[x+1][y].left = false;
            this.board.tiles[x][y+1].right = false;
            this.board.tiles[x+1][y+1].left = false;
            this.wallsleft-=1;
            if(!this.board.pawnsCanReachFinishLine()){
                this.removeWall(x,y);
                return false;
            }
            console.log('%d walls left', this.wallsleft);
            return true;
        }
        return false;
    }
    removeWall(x,y){
        if(x<0 || y<0 || x>7 || y>7){
            return false;
        }
        if(!this.board.corners[x][y].occupied){
            return false;
        }
        if(this.board.corners[x][y].wall.direction == HORIZONTAL){
            var owner = this.board.corners[x][y].wall.owner;
            owner.wallsleft += 1;
            this.board.corners[x][y].occupied = false;
            this.board.corners[x][y].wall = null;
            this.board.tiles[x][y].up = true;
            this.board.tiles[x+1][y].up = true;
            this.board.tiles[x][y+1].down = true;
            this.board.tiles[x+1][y+1].down = true;
            return true;
        }
        if(this.board.corners[x][y].wall.direction == VERTICAL){
            var owner = this.board.corners[x][y].wall.owner;
            owner.wallsleft += 1;
            this.board.corners[x][y].occupied = false;
            this.board.corners[x][y].wall = null;
            this.board.tiles[x][y].right = true;
            this.board.tiles[x+1][y].left = true;
            this.board.tiles[x][y+1].right = true;
            this.board.tiles[x+1][y+1].left = true;
            return true;
        }
        return false;
    }
}
class Board{
    constructor(){
        this.corners = [];
        for(var i=0;i<8;i++){
            this.corners.push([]);
            for(var j=0;j<8;j++){
                this.corners[i].push(new Corner(i,j));
            }
        }
        this.tiles = [];
        for(var i=0;i<9;i++){
            this.tiles.push([])
            for(var j=0;j<9;j++){
                var _tile = new Tile(i,j);
                if(i==0) _tile.left = false;
                if(i==8) _tile.right = false;
                if(j==0) _tile.down = false;
                if(j==8) _tile.up = false;
                if(i>0 && j>0) _tile.bottomleft = this.corners[i-1][j-1];
                if(i<8 && j>0) _tile.bottomright = this.corners[i][j-1];
                if(i>0 && j<8) _tile.topleft = this.corners[i-1][j];
                if(i<8 && j<8) _tile.topright = this.corners[i][j];
                this.tiles[i].push(_tile);
            }
        }
        this.pawn1 = new Pawn(0,this,4,0,10); //enemy
        this.pawn2 = new Pawn(1,this,4,8,10); //friendly
    }
    pawnsCanReachFinishLine(){
        return this.pawn1.canReachFinishLine() && this.pawn2.canReachFinishLine();
    }
    draw(){
        drawScreen();
        drawPawn(this.pawn1.x,this.pawn1.y,screen.colorset.enemy);
        drawPawn(this.pawn2.x,this.pawn2.y,screen.colorset.friendly);
        for(var i=0;i<8;i++){
            for(var j=0;j<8;j++){
                if(this.corners[i][j].occupied) drawWall(i,j,this.corners[i][j].wall.direction,screen.colorset.wall);
            }
        }
        drawRotationB();
    }
}